include/marl/dag.h

```diff
@@ -0,0 +1,390 @@
+  using NodeIndex = size_t;
+  using Work = typename detail::DAGWork<T>::type;
+  static const constexpr size_t NumReservedNodes = 32;
+  static const constexpr size_t NumReservedNumOuts = 4;
+  static const constexpr size_t InvalidCounterIndex = ~static_cast<size_t>(0);
+  static const constexpr NodeIndex RootIndex = 0;
+
+  // DAG work node.
+  struct Node {
+    MARL_NO_EXPORT inline Node() = default;
+    MARL_NO_EXPORT inline Node(Work&& work);
+
+    // The work to perform for this node in the graph.
+    Work work;
+
+    // counterIndex if valid, is the index of the counter in the RunContext for
+    // this node. The counter is decremented for each completed dependency task
+    // (ins), and once it reaches 0, this node will be invoked.
+    size_t counterIndex = InvalidCounterIndex;
+
+    // Indices for all downstream nodes.
+    containers::vector<NodeIndex, NumReservedNumOuts> outs;
+  };
+
+  // initCounters() allocates and initializes the ctx->coutners from
+  // initialCounters.
+  MARL_NO_EXPORT inline void initCounters(RunContext* ctx,
+                                          Allocator* allocator);
+
+  // notify() is called whenever a dependency task (ins) have completed.
+  // If all dependency tasks have completed (or this is the root node) then the
+  // node is run, and then notify() is called on all dependee nodes (outs).
+  MARL_NO_EXPORT inline void notify(RunContext*, NodeIndex);
+
+  // nodes is the full list of the nodes in the graph.
+  // nodes[0] is always the root node, which has no dependencies (ins).
+  containers::vector<Node, NumReservedNodes> nodes;
+
+  // initialCounters is a list of initial counter values to be copied to
+  // RunContext::counters on DAG<>::run().
+  // initialCounters is indexed by Node::counterIndex, and only contains counts
+  // for nodes that have at least 2 dependencies (ins) - because of this the
+  // number of entries in initialCounters may be fewer than nodes.
+  containers::vector<uint32_t, NumReservedNodes> initialCounters;
+};
+
+template <typename T>
+DAGBase<T>::Node::Node(Work&& work) : work(std::move(work)) {}
+
+template <typename T>
+void DAGBase<T>::initCounters(RunContext* ctx, Allocator* allocator) {
+  auto numCounters = initialCounters.size();
+  ctx->counters = allocator->make_unique_n<Counter>(numCounters);
+  for (size_t i = 0; i < numCounters; i++) {
+    ctx->counters.get()[i] = {initialCounters[i]};
+  }
+}
+
+template <typename T>
+void DAGBase<T>::notify(RunContext* ctx, NodeIndex nodeIdx) {
+  Node* node = &nodes[nodeIdx];
+
+  // If we have multiple dependencies, decrement the counter and check whether
+  // we've reached 0.
+  if (node->counterIndex != InvalidCounterIndex) {
+    auto counters = ctx->counters.get();
+    auto counter = --counters[node->counterIndex];
+    if (counter > 0) {
+      // Still waiting on more dependencies to finish.
+      return;
+    }
+  }
+
+  // Run this node's work.
+  if (node->work) {
+    ctx->invoke(node->work);
```

Comment:
what if the work enqueues tasks of its own? would the work just use a waitgroup/parallelize like normal (in which case invoke here wouldn't return until the join completed)? that seems pretty nice

Reply:
Yup. The work function is free to schedule new tasks and wait on them if it so wishes. The contract is simple: once the work function returns, the DAG node is complete and the dependees will execute as soon as possible. If you want to run sub DAGs from a work function, you can.